gitlab/data_source_gitlab_user.go

```diff
@@ -0,0 +1,131 @@
+package gitlab
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+func dataSourceGitlabUser() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceGitlabUserRead,
+
+		Schema: map[string]*schema.Schema{
+			"user_id": {
+				Type:          schema.TypeInt,
+				Computed:      true,
+				Optional:      true,
+				ConflictsWith: []string{"username"},
+			},
+			"username": {
+				Type:          schema.TypeString,
+				Computed:      true,
+				Optional:      true,
+				ConflictsWith: []string{"user_id"},
+			},
+			"email": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"is_admin": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"can_create_group": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"can_create_project": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"projects_limit": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"created_at": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"external": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"extern_uid": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"organization": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"two_factor_enabled": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceGitlabUserRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	var user *gitlab.User
+	var err error
+
+	userIdData, userIdOk := d.GetOk("user_id")
+	usernameData, usernameOk := d.GetOk("username")
+
+	// Get user by id
+	if userIdOk {
+		user, _, err = client.Users.GetUser(userIdData.(int))
+		if err != nil {
+			return err
+		}
+	} else {
+		// Get user by username
+		if !usernameOk {
+			return fmt.Errorf("one and only one of user_id or username must be set")
+		}
+
+		listUsersOptions := &gitlab.ListUsersOptions{}
+		username := usernameData.(string)
+		listUsersOptions.Username = &username
+
+		var users []*gitlab.User
+		users, _, err = client.Users.ListUsers(listUsersOptions)
+		if err != nil {
+			return err
+		}
+		if len(users) == 0 {
```

Comment:
I'd also check if `len(users) != 1` to report potential errors from the gitlab API.